src/lf_apply.rs

```diff
@@ -534,213 +541,171 @@ unsafe fn filter_plane_rows_uv<BD: BitDepth>(
-                    let ref mut fresh13 = (*uv_vmask
-                        .offset(cmp::min(idx_2, (*a).tx_lpf_uv[i_0 as usize] as c_int) as isize))
-                        [sidx_2 as usize];
-                    *fresh13 = (*fresh13 as c_uint | smask_2) as u16;
-                    uv_mask_0 <<= 1;
-                    i_0 = i_0.wrapping_add(1);
+                let uv_vmask: &mut [[u16; 2]; 2] =
+                    &mut (*lflvl.offset(x as isize)).filter_uv[1][(starty4 >> ss_ver) as usize];
+                for i in 0..cw {
+                    let uv_mask: u32 = 1 << i;
+                    let sidx = (uv_mask >= hmax) as usize;
+                    let smask = (uv_mask >> (sidx << 4 - ss_hor)) as u16;
+                    let idx = (uv_vmask[1][sidx] & smask != 0) as usize;
+                    uv_vmask[1][sidx] &= !smask;
+                    uv_vmask[0][sidx] &= !smask;
+                    uv_vmask[cmp::min(idx, a[0].tx_lpf_uv[i as usize] as usize)][sidx] |= smask;
                 }
             }
-            x += 1;
-            a = a.offset(1);
+            a = &a[1..];
         }
     }
-    let mut ptr: *mut BD::Pixel;
     let level_ptr = &(*f).lf.level[((*f).b4_stride * sby as isize * sbsz as isize) as usize..];
-    ptr = p[0];
-    have_left = 0 as c_int;
+    let mut offset = p_offset[0];
+    have_left = false;
     for (x, level_ptr) in (0..(*f).sb128w).zip(level_ptr.chunks(32)) {
         filter_plane_cols_y::<BD>(
             f,
             have_left,
             level_ptr,
             (*f).b4_stride,
-            ((*lflvl.offset(x as isize)).filter_y[0]).as_mut_ptr() as *const [[u16; 2]; 3],
-            ptr,
+            &(*lflvl.offset(x as isize)).filter_y[0],
+            p[0],
+            offset,
             (*f).cur.stride[0],
-            cmp::min(32 as c_int, (*f).w4 - x * 32),
-            starty4,
+            cmp::min(32, (*f).w4 - x * 32),
+            starty4 as c_int,
             endy4 as c_int,
         );
-        have_left = 1 as c_int;
-        ptr = ptr.offset(128);
+        have_left = true;
+        offset += 128;
     }
     if frame_hdr.loopfilter.level_u == 0 && frame_hdr.loopfilter.level_v == 0 {
         return;
     }
-    let mut uv_off: ptrdiff_t;
     let level_ptr = &(*f).lf.level[((*f).b4_stride * (sby * sbsz >> ss_ver) as isize) as usize..];
-    have_left = 0 as c_int;
-    uv_off = 0;
+    let (pu, pv) = p[1..].split_at_mut(1);
+    let mut uv_off = p_offset[1];
+    have_left = false;
     for (x, level_ptr) in (0..(*f).sb128w).zip(level_ptr.chunks(32 >> ss_hor)) {
         filter_plane_cols_uv::<BD>(
             f,
             have_left,
             level_ptr,
             (*f).b4_stride,
-            ((*lflvl.offset(x as isize)).filter_uv[0]).as_mut_ptr() as *const [[u16; 2]; 2],
-            &mut *p[1].offset(uv_off as isize),
-            &mut *p[2].offset(uv_off as isize),
+            &(*lflvl.offset(x as isize)).filter_uv[0],
+            pu[0],
+            pv[0],
```

Comment:
```suggestion
            p[1],
            p[2],
```

Why can't you just do this?

Reply:
Can't borrow `p[1]` and `p[2]`.

Reply:
It seemed to work fine when I tried changing it to just use `p` directly.

Reply:
My bad, you're right.  You can do it more elegantly, though:

```suggestion
            pu,
            pv,
```

with a `let [_, pu, pv] = p;` array destructure above instead.

Reply:
That's much more elegant, indeed.